src/allocator_defrag.c

```diff
@@ -0,0 +1,379 @@
+} jeBinInfoKeys;
+
+/* Represents detailed information about a jemalloc bin.
+ *
+ * This struct provides metadata about a jemalloc bin, including the size of
+ * its regions, total number of regions, and related MIB keys for efficient
+ * queries.
+ */
+typedef struct jeBinInfo {
+    unsigned long reg_size;  /* Size of each region in the bin. */
+    unsigned long nregs;     /* Total number of regions in the bin. */
+    jeBinInfoKeys info_keys; /* Precomputed MIB keys for querying bin statistics. */
+} jeBinInfo;
+
+/* Represents the configuration for jemalloc bins.
+ *
+ * This struct contains information about the number of bins and metadata for
+ * each bin, as well as precomputed keys for batch utility queries and epoch updates.
+ */
+typedef struct jeBinsConf {
+    unsigned long nbins;           /* Number of bins in the jemalloc configuration. */
+    jeBinInfo *bin_info;           /* Array of `jeBinInfo` structs, one for each bin. */
+    jeMallctlKey util_batch_query; /* Key to query batch utilization information. */
+    jeMallctlKey epoch;            /* Key to trigger statistics sync between threads. */
+} jeBinsConf;
+
+/* Represents the latest usage statistics for a jemalloc bin.
+ *
+ * This struct tracks the current usage of a bin, including the number of slabs
+ * and regions, and calculates the number of full slabs from other fields.
+ */
+typedef struct jeBusage {
+    unsigned long curr_slabs;         /* Current number of slabs in the bin. */
+    unsigned long curr_nonfull_slabs; /* Current number of non-full slabs in the bin. */
+    unsigned long curr_full_slabs;    /* Current number of full slabs in the bin (calculated from other fields). */
+    unsigned long curr_regs;          /* Current number of regions in the bin. */
+} jeBusage;
+
+
+static int defrag_supported = 0;
+static size_t jemalloc_quantum = 0;
+static jeBinsConf arena_bin_conf = {0, NULL, {{0}, 0}, {{0}, 0}};
+static jeBusage *curr_usage = NULL;
+
+
+/* -----------------------------------------------------------------------------
+ * Alloc/Free API that are cooperative with defrag
+ * -------------------------------------------------------------------------- */
+
+/* Allocation and free functions that bypass the thread cache
+ * and go straight to the allocator arena bins.
+ * Currently implemented only for jemalloc. Used for online defragmentation.
+ */
+void *allocatorDefragAlloc(size_t size) {
+    void *ptr = je_mallocx(size, MALLOCX_TCACHE_NONE);
+    return ptr;
+}
+void allocatorDefragFree(void *ptr, size_t size) {
+    if (ptr == NULL) return;
+    je_sdallocx(ptr, size, MALLOCX_TCACHE_NONE);
+}
+
+/* -----------------------------------------------------------------------------
+ * Helper functions for jemalloc translation between size and index
+ * -------------------------------------------------------------------------- */
+#define LG_QUANTOM_8_FIRST_POW2 3
```

Comment:
Quantum 8 = lg-quantum 3, right? Should this be `LG_QUANTUM_3_FIRST_POW2`?

Btw, quantum is misspelled.

Reply:
yes, will fix